Core/Src/it.c

```diff
@@ -6,26 +6,43 @@
  */
 
 #include "system.h"
+//#include "communicator.h"
 
 extern System my_sys;
+extern UART_HandleTypeDef huart3;
+
 void HAL_TIM_PeriodElapsedCallback(TIM_HandleTypeDef *htim) {
 	if (htim == (&my_sys)->LED1_timer) {
 		(&my_sys)->LEDToggle1 = 1; // flip flag
 	}
 	if (htim == (&my_sys)->LED2_timer) {
 		(&my_sys)->LEDToggle2 = 1; // flip flag
 	}
+	if (htim == (&my_sys)->handshakeTimer) {
+		my_sys.receive_ring_buffer->push(my_sys.receive_ring_buffer, HANDSHAKE_ARE_YOU_THERE);
+		my_sys.LED1_State = OFF;
+		my_sys.LED2_State = OFF;
+	}
 }
 
+void HAL_UART_RxCpltCallback(UART_HandleTypeDef *huart) {
 
-/*
-void HAL_UART_RxCpltCallback(UART_HandleTypeDef *huart)
-{
-   Prevent unused argument(s) compilation warning
-  UNUSED(huart);
+	if (huart==&huart3) {
+		my_sys.handshakeTimer->Instance->CNT=0;
+//		my_sys.received_input = TRUE;
+		my_sys.receive_ring_buffer->push(my_sys.receive_ring_buffer, *my_sys.incomingByte);
+
+
+	}
 
-   NOTE : This function should not be modified, when the callback is needed,
-            the HAL_UART_RxCpltCallback can be implemented in the user file.
+}
+
+
+void HAL_UART_TxCpltCallback(UART_HandleTypeDef *huart) {
+
+	if (huart==&huart3) {
+//		my_sys.received_input = FALSE;
```

Comment:
I notice that you actually do send more Tx interrupts than just the "are you there", so I would consider using the huart3 handle's state variable in an infinite while loop spin lock.  Added points if you also give it a timeout so that you don't kill the program on unknown edge cases.
`while(huart->gState != HAL_UART_STATE_READY) {}
HAL_UART_Transmit_IT();`
Or just not using a transmit interrupt is the alternative

Reply:
I'm pretty sure this is already built into HAL_UART_Transmit_IT

Reply:
The main thing HAL_UART_Transmit_IT doesn't do is have a timeout.  If it tries to utilize the huart3 peripheral and it is already occupied, it will simply not do anything and return an error.  So if the last Transmit call is still underway when the next one is called (since we are using non-blocking operations), the program will fail.
This could happen if a user inputs a character _exactly_ as the 10 second "are you still there" timer gets executed, so you will either need to have some sort of wait/delay while loop for the peripheral to become available again, or simply not use interrupts for transmitting data.

Reply:
Unless I'm missing something, it looks like you didn't modify either the callback function or the interrupt TX calls.  Is there a change somehwere I am missing?